BUG/PERF: Stata value labels

kshedden opened the pull request:
closes #12014 

This PR fixes a minor bug and introduces some performance enhancements, all related to value label reading in Stata files.  

The bug is that when reading a Stata file incrementally, the value labels will be read even when specifying convert_categoricals=False (this does not happen when reading the entire file at once).

The performance enhancements are:
1. Read key/value information as an ndarray using np.frombuffer rather than as a Python loop.
2. When splitting the value labels at the offsets, we currently pass txt[off[i]:] to null_terminate, which creates many copies of large segments of a potentially large byte array.  I changed it so that
   only the relevant part of the string is copied.

Relating to 2, further performance improvements might be possible since there is no trailing null byte to remove except for the last element of `txt` (thus some of the work in `_null_terminate` is superfluous).

Background:  This is an issue when processing large Stata files with millions of distinct value labels.

kshedden committed:
Initial commit for PR

added text to whatsnew

Update whatsnew

flake8 edits

edited whatsnew

jreback commented:
ok, sounds gr8!.

just confirm that we have reasonable benchmarks in ``asv`` (the ``packers.py``) file is the only one for stata.

kshedden commented:
Is there a place to put data files for ASV to read in?  The stata writer is more limited than the stata reader, so it's hard to test some features like strls that are not part of dta version 114.

jreback commented:
you can just make a directory under the ``asv_benchmarks``

jreback commented:
@kshedden want to update

kshedden commented:
I added a test but haven't been able to get asv to run it.  I don't have conda so switched the asv conf file to use virtualenv.  It fails when using pip to install pytables:

```(py3)-bash-4.1$ pip install --upgrade pytables
Collecting pytables
  Could not find a version that satisfies the requirement pytables (from versions: )
No matching distribution found for pytables```

I'm using 3.4.2rc1 (and also switched asv conf to use python 3.4)

jreback commented:
asv *can* work with pip but needs a slighty modified config file (e.g. a different one) as pip uses ``tables`` (not that *conda* accepts both of these).

jreback commented:
pls add a whatsnew note (in performance). you don't necessarily need to include an asv benchmark (though nice if its easy), but post a perf comparison at the top of the issue.

pv commented:
Re conda vs. virtualenv, this may be of interest: https://github.com/spacetelescope/asv/issues/322#issuecomment-150848137 https://github.com/spacetelescope/asv/pull/329

jreback commented:
@pv that's a nice feature....care to do a PR to upgrade ``pandas/asv.conf.json``

jreback commented:
@kshedden can you update

jreback commented:
@kshedden can you update

kshedden commented:
Sorry for the delay... I wasn't able to test this in ASV.  I updated whatsnew, not sure what else needs to be done.

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -416,7 +416,7 @@ Performance Improvements
 
 - Improved performance of ``andrews_curves`` (:issue:`11534`)
-
+- Improved performance of ``StataReader``
```
add this issue number here

jreback commented:
looks good
can u run flake8 on the diff and fix any issue

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -440,6 +439,7 @@ Bug Fixes
 - Bug in consistency of passing nested dicts to ``.groupby(...).agg(...)`` (:issue:`9052`)
 - Accept unicode in ``Timedelta`` constructor (:issue:`11995`)
 
+- Bug in value label reading for ``StataReader`` when reading incrementally (:issue:`12014`)
```
this is also a perf enhancement, yes? pls add a line in Performance (use this PR number for it)

jreback commented:
minor comments. pls squash. ping when green.

kshedden commented:
@jreback should be good to go

jreback commented:
merged via https://github.com/pydata/pandas/commit/449ab6b6646812c21e859f71d4aabda2288e8ec4

thanks!
